### PR TITLE
Emphasize relative importance and harden + consolidate "Effects of Priority Hints"

### LINKED ISSUE #32: Spec should specify (and emphasize) "relative", importance
At the Web Perf WG call today, it seemed that one of @toddreifsteck's more prominent hesitations revolved around the notion that Priority Hints might be trying to shift priorities among the many and different types (destinations, really) of resources. For example, a concern would have been "Well would a high importance image get prioritized higher than a low importance CSS file? And what if it did in Chrome but not Edge? etc." This sounded a bit like the original intent of Priority Hints (I recall group names and https://github.com/WICG/priority-hints/issues/7#issuecomment-331070080, which seems solved by specifying and emphasizing _relative_ importance). I think a breakthrough we had in the conversation was when we mentioned that the prioritization of resources should be relative between types/destinations. For example, no matter what prioritization scheme might be used in Edge vs Chrome vs FF vs Safari, low-importance images should be prioritized lower than *other non-low-importance* images. The same goes for scripts, fetch()es, etc. Modifying the priority of a request relative to other requests of the same type definitely seems like something we'd like to make clear(er).

Assigning this to Addy right now as he's been digging around the spec quite a bit lately, but happy to pick this up if needbe after some more implementation work.

----

(aside from the main point of this issue:)

Another very reasonable hesitation we got revolved around the fact that it might be helpful to fully specify exactly how a resource gets prioritized, and some of the surrounding infrastructure, values, and logic so we could have this work dependably between the different implementations. While this is very valuable, I (and I believe @yoavweiss) think that Priority Hints would actually be a good step to getting us there, and is not a blocker so much as a catalyst to that conversation (which there has been some, albeit not tons, of talk around). For example, by talk about how `importance=high` async scripts should be prioritized higher than `importance=low` async scripts, we're already making relative comparisons that the Fetch Standard does not yet talk about.

### PR DESCRIPTION
**Pull request by domfarolino**
This PR attempts to fix #32. A few things it does:

 - Specifies exactly what the expected effects of Priority Hints are in a new section, since it is possible to see the current state of this as vague and under-specified.
 - Includes an example showing that Priority Hints are meant to be a relative influencer, ideally clearing up some confusion in https://github.com/mozilla/standards-positions/issues/25 and elsewhere.
- Attempts to consolidate references to what exactly the expected effects of the hints are. There are some cases where we reference things like "...affect the HTTP/2 weights and dependencies of a request" from several locations. I attempt to cut a few of these out in favor of referencing the newly added section. This might make the document more maintainable.
 - Adds more definition dependencies that I refer to in this PR + the HTML Standard biblio entry
 - Changes a few instances of "may" to "might", in an effort to use more non-normative language where need-be (please see https://infra.spec.whatwg.org/#conformance)

/cc @kenjibaheux @kinu @slightlyoff @toddreifsteck @wanderview

----

With some of the newer changes to the spec, the Explainer doc is getting more and more out of date. Have we looked into what we do with the explainer once the spec gets more mature? Remove it? Delete it?

### CODE REVIEW

**Comment by KenjiBaheux** [comment]
I've read the proposed changes and they seemed reasonable to me.
But I defer to Addy for the approval.

**Review by addyosmani — APPROVED** [review]
LGTM.

File: index.html [hunk 1]
```diff
@@ -81,79 +82,135 @@ <h2>Dependencies</h2>
...
-      to their heuristic priority. Browsers may also use this heuristic resource priority to delay sending certain requests
+      to their heuristic priority. Browsers might also use this heuristic resource priority to delay sending certain requests
       in order to avoid bandwidth contention of these resources with more critical ones.</p>
 
     <p>Currently web developers have very little control over the heuristic importance of loaded resources, other than speeding
-      up their discovery using
-      <code>&lt;link rel=preload&gt;</code>. Browsers make many assumptions on the importance of resources based on the resource's type (AKA its request destination),
-      and its location in the containing document.</p>
+      up their discovery using <code>&lt;link rel=preload&gt;</code>. Browsers mostly determine a request's priority based on the
+      resource's <a>destination</a>, and location in the containing document if applicable.</p>
```
Comment (yoavweiss) [inline index.html]: destination is a property of the request, no?

File: index.html [hunk 2]
```diff
@@ -81,79 +82,135 @@ <h2>Dependencies</h2>
...
 
-    <p>This document will detail use cases and a markup sketch that will provide developers with the control to indicate a resource's
-      relative importance to the browser, and enable the browser to act on those indications to modify the time it sends
-      out the request and its HTTP/2 dependency/weight so that important resources are fetched and used earlier, in order
-      to improve the user's experience.</p>
+    <p>This document details use cases and a markup sketch that will provide developers with the control to indicate a resource's
```
Comment (yoavweiss) [inline index.html]: I think we can remove the "sketch" part

File: index.html [hunk 3]
```diff
@@ -81,79 +82,135 @@ <h2>Dependencies</h2>
...
       resource when fetching it. Note that this is intentionally weak language, allowing for a browser to apply its own preferences
-      for resource priority or heuristics if deemed important.
+      for resource priority or heuristics if deemed important. See the below section for more information.
     </p>
+
     <p>
       <div class="note">
-        <p>Priority Hints compliment existing browser loading primitives such as preload. Preload is a mandatory and high-priority
-          fetch for a resource that is necessary for the current navigation. Priority Hints can hint that a resource's priority should be lower or higher than its default, and can also be used to provide more granular prioritization to preloads.</p>
+        <p>Priority Hints compliment existing browser loading primitives such as preload. Preload is a mandatory fetch for a resource
```
Comment (yoavweiss) [inline index.html]: s/compliment/complement/

File: index.html [hunk 4]
```diff
@@ -81,79 +82,135 @@ <h2>Dependencies</h2>
...
+          its default, and can also be used to provide more granular prioritization to preloads.</p>
       </div>
     </p>
+
     <p>This is how we conceptually think about different resource types under the hood in browsers today. It may translate well
       to user-space where different types of content share similar properties.</p>
 
+    <h2 id="effects-of-priority-hints">Effects of Priority Hints</h2>
+
+    <p>Exactly what implementations do with Priority Hints is up to them since much of the resource loading pipeline is currently
```
Comment (yoavweiss) [inline index.html]: Maybe rephrase? e.g. "Since user agents' resource loading pipeline is largely unspecified, this document doesn't mandate specific mapping between priority hints and request prioritization. However, it does describe some ways in which..."

File: index.html [hunk 5]
```diff
@@ -81,79 +82,135 @@ <h2>Dependencies</h2>
...
+        If requests for <code>image</code>
+        <a data-lt="destination">destinations</a> in a particular implementation are typically assigned a stream weight of
+        <code>60</code>, a request for an image with <code>importance="low"</code> might be assigned a stream weight less than
+        <code>60</code>. In other words, <code>importance="low"</code> on an image might lead to an entirely different resolved
+        HTTP/2 stream priority than <code>importance="low"</code> on something like a script, or an iframe.
+      </p>
+    </div>
+
+    <strong>Queueing</strong>
+    <p>A browser might choose to queue up certain low priority requests until higher priority requests are sent out or finished
```
Comment (yoavweiss) [inline index.html]: s/browser/user agent/

**Review by yoavweiss — COMMENTED** [review]
Thanks for working on this! :) A few comments

**Comment by yoavweiss** [comment]
@toddreifsteck - can you take a look and say if this addresses your concerns?

**Comment by domfarolino** [comment]
Thanks for the review everyone! I've addressed Yoav's comments. It'd be great if Todd could take a look and share thoughts, but won't I block merging in the mean time.
